### PR TITLE
OpenAIRE guidelines for Data Archives

### PR DESCRIPTION
**Pull request by abollini**
This PR introduce compliancy with the OpenAIRE guidelines for Data Archives, see
https://guidelines.openaire.eu/en/latest/data/index.html
as requested in the issue #4257

It is target to the new version of the guidelines the 4.0 not yet published but it was anticipated that the main change will be the switch to the datacite schema 4.1

This development has been provided by 4Science thanks to the funding from the OpenAIRE Open Call for Innovation Jan 2018 see https://www.openaire.eu/openaire-tender-calls-winners

As we doesn't currently host installation with real data it will be useful to have some feedback and test from Insititution with data already suitable for exposure to openaire, better if with reference to EC funded projects.

---

Update by @pdurbin on 2018-05-14

connects to #4257 As an installation admin, I want my repository to export OpenAIRE-compliant metadata to improve discoverability, reusability of research data

I added the "connects to" line above to associate this pull request with the issue it addresses in Waffle ( https://waffle.io/IQSS/dataverse ).

---

Update by @pdurbin on 2018-05-17

connects to #3697 Users should be able to download DataCite metadata from Export Metadata pulldown

### CODE REVIEW

**Comment by pdurbin** [comment]
@abollini thanks for the pull request! The first think I noticed is that the Travis build failed with a compilation error. If you click "Details" ...

[FIXME]

you will be taken to https://travis-ci.org/IQSS/dataverse/builds/378185244 and if you scroll down you can see the error:

[FIXME]

I'm sorry but I can't seem to upload images from the airport WiFi in Qatar so I put FIXME above for now.

Can you please take a look at the tests that are failing. We run the test suite on every build. I'm not sure if this is a Netbeans default. You seem to be using Eclipse.

**Comment by pameyer** [comment]
06ce109df7d6b6495eb5bebafc718215d6a883ab is ok for integration tests: ``Tests run: 72, Failures: 0, Errors: 0, Skipped: 0``

**Comment by pdurbin** [comment]
@juancorr thanks for testing. If @abollini gives you push access to his branch, are you able to work on these issues?

**Comment by djbrooke** [comment]
ping @abollini - I think the feedback above is for you. Thanks @juancorr for the feedback!

**Comment by juancorr** [comment]
Thanks @pdurbin,  yes I can work until 24 Mai, I will out of the office two of three weeks after this date.

**Comment by pdurbin** [comment]
@juancorr ok, and you could always push a branch to your own fork and make a pull request from your fork to @abollini 's fork. That way he wouldn't need to give you push access.

**Comment by abollini** [comment]
Hi @pdurbin , @juancorr thanks for the feedback, sorry for the delay in the response but I was travelling in the past days. You should be able to add commit yourself as we keep flagged the "Allow edits from maintainers." flag, anyway we can make the chages ourself on Monday if you don't have done them already!

Thanks again for the fast feedback

**Comment by pdurbin** [comment]
@abollini no problem and thanks again for the pull request. If you could make the changes, it would be great. @juancorr is not a maintainer so he won't be able to push to your branch unless you give him special access. I'll let you two coordinate how you'd like to make further improvements to the code before we advance it to code review at https://waffle.io/IQSS/dataverse . Please just let us know when you're ready. One other bit of feedback I'd like to give is that it would be nice if there were JUnit tests for `OpenAireExporter.java`. If this is too much trouble, don't worry about it, but you could look at `SchemaDotOrgExporterTest.java` as an example.

(Thank you to @francescopioscognamiglio for fixing the compilation error in 06ce109 !)

**Comment by juancorr** [comment]
Thanks @abollini and @pdurbin, I can have the code ready Monday or Tuesday

**Comment by pdurbin** [comment]
No rush but please note that there are now merge conflicts because pull request #4691 was merged.

**Comment by lap82** [comment]
As said at https://github.com/4Science/dataverse/pull/2 we working on cleaning code merged from @juancorr and we will try to introduce the publisher (as @juancorr pointed out that "Publisher is mandatory in xsd file (not in guidelines)"). We will share asap the code in this PR. Thank you!

**Comment by pdurbin** [comment]
@lap82 thanks for the update and for pointing out https://github.com/4Science/dataverse/pull/2 to me. Sounds like great progress! Thanks!

**Comment by pdurbin** [comment]
@francescopioscognamiglio thanks! It sounds like you're ready for code review so I just dragged this pull request (and #4257 and #3697) to that column at https://waffle.io/IQSS/dataverse

**Comment by djbrooke** [comment]
Hi @abollini @lap82 and @francescopioscognamiglio (and anyone else interested) - we'll soon start taking a look at the code, but can you add some guides documentation as well? If you're unable to update the .rst files directly, a document that we can use to update the documentation would be fine as well! Thanks for the work on this!

**Comment by pdurbin** [comment]
Specifically, adding the new export format to this file would be a good start: `doc/sphinx-guides/source/admin/metadataexport.rst`

**Comment by pameyer** [comment]
b6743d80c95333591ea8366eb6417ca2f1d9866d appears to have a failing integration test.

`Tests run: 72, Failures: 1, Errors: 0, Skipped: 0`

`Failed tests:   testDatasetThumbnail(edu.harvard.iq.dataverse.api.SearchIT): JSON path data.datasetThumbnailBase64image doesn't match.(..)`

It does look like SearchIT has changed in this PR, suggesting that it's a change in behavior.

**Comment by pdurbin** [comment]
@pameyer thanks for discovering that! @abollini or @francescopioscognamiglio or @juancorr would you like to try to reproduce the failing API test? The easiest way is probably to use the "All In One Docker Images for Testing" described at http://guides.dataverse.org/en/4.8.6/developers/containers.html#all-in-one-docker-images-for-testing

**Comment by pameyer** [comment]
Just checked develop, and as of d6850914b8c91aa468a65d9937114ffb49199b6b the same failure is present there - so it appears unrelated to this PR.

**Comment by pdurbin** [comment]
Ah, thanks, @pameyer . I'll go kick off a build over at https://build.hmdc.harvard.edu:8443/job/phoenix.dataverse.org-apitest-develop/

**Comment by pdurbin** [comment]
Yeah, I'm sure the SearchIT failure has nothing to do with this pull request. We're tracking it at #4719. Thanks, @pameyer

**Comment by djbrooke** [comment]
Thanks @pdurbin for shepherding this and for the quick discussion over scavenged pizza just now. Feel free to add the documentation as we discussed and then this is ready for Code Review. 

 @abollini @lap82 and @francescopioscognamiglio - some tests would be great here if you have some cycles, but @pdurbin did mention that it was a challenging area to add automated tests.

**Comment by jggautier** [comment]
I've been leaving comments about this PR in the connected issue #4257, but as I keep reviewing the metadata, I realize now it's better if I leave comments about the metadata mapping in PR comments here, so they're more associated with the PR as it changes.

Here are the changes I've noticed since I looked at the xml metadata records that this PR makes harvestable, and my remaining questions:

- I couldn't see the option in the Export Metadata pulldown to export OpenAIRE DataCite metadata. That was a bonus for the scope of this issue (#4257), so it wasn't necessary for the goal of making Dataverse OpenAIRE compatible, but I just wanted to point out this change.

- If any file in the dataset is restricted, the rights property is now "restrictedAccess," whether or not "request access" is enabled:

  screen shot 2018-07-17 at 3 27 21 pm

  This seems better to me than using closedAccess.

- Could the xml:lang attributes not be used for now? I'm suggesting this for two reasons:

  a. Dataverse's "Language" field and OpenAIRE's "language" property should define the language of the resource (the dataset). But here the Dataverse language field is being used to populate the xml language attributes of certain OpenAIRE fields:

  <title xml:lang="English">Historical Climate Model Output Of Echam5-Wiso From 1871-2011 At T106 Resolution</title>

  When a Dataverse depositor chooses French, she's saying that the dataset is in English, and not necessarily that the metadata she's entering is in French. But in this example, this PR uses French to describe the language of the metadata as well, which isn't always true.

  b. DataCite is checking what's in the ```xml:lang``` attribute against some list of languages (an ISO list?). So the xml won't be valid if a dataset depositor chooses any of dozens of languages in Dataverse's language vocabulary that aren't on that list. Ideally, Dataverse would include ISO codes for its languages vocab. I propose that the community consider adding ISO language codes at another time (in another issue), and for now, the xml:lang attribute not be used in the metadata that OpenAIRE harvests.

- It looks like the rules for how the NameType attribute is used has changed (or I didn't understand them the first time I looked):

  If the name has a comma, then the nametype is Personal if there's also an affiliation and/or an ID/ ID Scheme. (Neither ID nor ID Scheme will be in the OpenAIRE DataCite metadata unless both have values.)

  If the name has no comma, then there is no nametype unless there is an ID and ID Scheme.

  This seems reasonable to me (until the community can consider other ways (such as https://github.com/IQSS/dataverse/issues/2243#issuecomment-358625744) of figuring out if the author is a person or organization).

- If "Funder" is chosen as the contributor type, could the Contributor Name be mapped to Funder name? There is no "Funder" contributor type in DataCite 4.1. (I opened a new issue, outside of the scope of this one, for removing the "Funder" contributor type, https://github.com/IQSS/dataverse/issues/4859)

Discussion about how to test that this PR lets OpenAIRE harvest from Dataverse is in the [in the Google Groups thread](https://groups.google.com/d/msg/dataverse-community/OALTzINxkX0/vx0T9CVADAAJ)).

**Comment by pdurbin** [comment]
This pull request now has merge conflicts.

**Comment by fcadili** [comment]
An update version of code was committed. 

The following elements was improved:

- Export Metadata works, see sample at http://ec2-100-27-31-230.compute-1.amazonaws.com:8080/dataset.xhtml?persistentId=doi:10.5072/FK2/G251YB

- property "restrictedAccess," is handle whether or not "request access" is enabled. The required behaviour is checked in tests testWriteRestrictedAccessRightElementWithRequestAccessDisabled and testWriteRestrictedAccessRightElementWithRequestAccessEnabled.

- xml:lang attributes will not be used for now. The getLanguage method is no longer called to initialize language variable.

- "Funder" is mapped to funder name. The Funder is removed from contributors and added to fundingReferences. The required behaviour is checked in tests testWriteFunderNamePropertyNotInContributor and testWriteFunderNamePropertyInFundingReferencesElement.

- xml validation of full document is done in test testValidateExportDataset.

- test coverage was improved on single methods of OpenAireExportUtil and adding xml validation test.

**Comment by fcadili** [comment]
An update version of code was committed.

The following elements was improved:
- DataCite algorithm described in #2243 (comment) was applied to handle creator nametype properly.

File: src/main/java/edu/harvard/iq/dataverse/export/openaire/OpenAireExportUtil.java [hunk 1]
```diff
@@ -0,0 +1,1518 @@
...
+            if ("citation".equals(key)) {
+                for (FieldDTO fieldDTO : value.getFields()) {
+                    if (DatasetFieldConstant.author.equals(fieldDTO.getTypeName())) {
+                        for (HashSet<FieldDTO> foo : fieldDTO.getMultipleCompound()) {
+                            String creatorName = null;
+                            String affiliation = null;
+                            String nameIdentifier = null;
+                            String nameIdentifierScheme = null;
+
+                            for (Iterator<FieldDTO> iterator = foo.iterator(); iterator.hasNext();) {
```
Comment (pdurbin) [inline src/main/java/edu/harvard/iq/dataverse/export/openaire/OpenAireExportUtil.java]: All of these "foo" variables were copied from DublinCoreExportUtil, I assume.

**Comment by fcadili** [comment]
I see your examples and I will correct the contributor's name type to handle the proposed case as a non-personal nameType. Instead for the example relative to creator's name type I think it is correct as described in [#2243](https://github.com/IQSS/dataverse/issues/2243) (comment)

**Comment by pdurbin** [comment]
> I see your examples and I will correct the contributor's name type to handle the proposed case as a non-personal nameType

@fcadili hi! I'm not too worried about the Guido van Rossum example I added in 03c959b if that's what you mean.

Much more important than anything I wrote in my review of this pull request is the feedback from QA at https://github.com/IQSS/dataverse/issues/4257#issuecomment-485863609 . I just want to make sure you've seen that comment. When you're ready, please reply on the issue. Thanks!!

**Comment by fcadili** [comment]
An update version of code was committed.

The following elements was improved:
- Organizational value of contributor's nameType is used when contributorName is not a person.

**Comment by fcadili** [comment]
I posted an updated version of the code that resolve @kcondon questions.

nameType attribute is still used, but it can be removed when needed. 
However I note when we replace:
```
<creators>
    <creator>
      <creatorName nameType="Personal">Finch, Fiona</creatorName>
      <givenName>Fiona</givenName>
      <familyName>Finch</familyName>
      <nameIdentifier nameIdentifierScheme="ORCID">0000-0002-9528-9470</nameIdentifier>
      <affiliation>Birds Inc.</affiliation>
    </creator>
  </creators>
```
with
```
<creators>
    <creator>
      <creatorName>Finch, Fiona</creatorName>
      <givenName>Fiona</givenName>
      <familyName>Finch</familyName>
      <nameIdentifier nameIdentifierScheme="ORCID">0000-0002-9528-9470</nameIdentifier>
      <affiliation>Birds Inc.</affiliation>
    </creator>
  </creators>
```
we sill need to establish that creatorName is a person, since elements givenName and familyName have to be generated.

**Comment by pdurbin** [comment]
@fcadili hmm. Good point about those separate fields for first and last name. Is there a database setting to control the loadHint boolean? If not, how does a sysadmin change it?

**Comment by fcadili** [comment]
At the moment there is no database field. However it can be done with some help to identify database table and column and method used to change the value (using api or direct database modification).

**Comment by fcadili** [comment]
I released a new version: the problem was related to resource files that was not deployed with dataverse.war. I checked all @kcondon examples and all seems work fine.

**Comment by fcadili** [comment]
I released a new version which fixed Organizational nameType. I've checked all @kcondon examples and they works fine with the new version.

**Comment by fcadili** [comment]
I released a new version which fixes Operational nameType when the name contains commas or dashes. I've checked the latest @kcondon examples and they work fine with the new version.

**Pull request merged**